### PR TITLE
[SPARK-48706][PYTHON] Python UDF in higher order functions should not throw internal error

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR fixes the error messages and classes when Python UDFs are used in higher order functions.

### Why are the changes needed?

To show the proper user-facing exceptions with error classes.

### Does this PR introduce _any_ user-facing change?

Yes, previously it threw internal error such as:

```python
from pyspark.sql.functions import transform, udf, col, array
spark.range(1).select(transform(array("id"), lambda x: udf(lambda y: y)(x))).collect()
```

Before:

```
py4j.protocol.Py4JJavaError: An error occurred while calling o74.collectToPython.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 15 in stage 0.0 failed 1 times, most recent failure: Lost task 15.0 in stage 0.0 (TID 15) (ip-192-168-123-103.ap-northeast-2.compute.internal executor driver): org.apache.spark.SparkException: [INTERNAL_ERROR] Cannot evaluate expression: <lambda>(lambda x_0#3L)#2 SQLSTATE: XX000
	at org.apache.spark.SparkException$.internalError(SparkException.scala:92)
	at org.apache.spark.SparkException$.internalError(SparkException.scala:96)
```

After:

```
pyspark.errors.exceptions.captured.AnalysisException: [INVALID_LAMBDA_FUNCTION_CALL.UNEVALUABLE] Invalid lambda function call. Python UDFs should be used in a lambda function at a higher order function. However, "<lambda>(lambda x_0#3L)" was a Python UDF. SQLSTATE: 42K0D;
Project [transform(array(id#0L), lambdafunction(<lambda>(lambda x_0#3L)#2, lambda x_0#3L, false)) AS transform(array(id), lambdafunction(<lambda>(lambda x_0#3L), namedlambdavariable()))#4]
+- Range (0, 1, step=1, splits=Some(16))
```

### How was this patch tested?

Unittest was added

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @cloud-fan and @ueshin

**Comment by cloud-fan** [comment]
This is more like an unsupported feature, that we don't support Python UDF as the lambda function within a higher order function. How about the `UNSUPPORTED_FEATURE` error?

BTW, will we ever support it in the future?

**Comment by HyukjinKwon** [comment]
I don't think we will ever support this :-). Let me fix up the error

File: common/utils/src/main/resources/error/error-conditions.json [hunk 1]
```diff
@@ -4645,6 +4645,11 @@
         "message" : [
           "TRANSFORM with SERDE is only supported in hive mode."
         ]
+      },
+      "LAMBDA_FUNCTION_WITH_PYTHON_UDF" : {
+        "message" : [
+          "Python UDFs with a lambda function at a higher order function is unsupported. However, a Python UDF <funcName> was found."
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-conditions.json:4651]: The full error message starts with `The feature is not supported: `, so how about `Python UDF <name> within a lambda function.`

Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-conditions.json:4651]: or `Lambda function with Python UDF <name>`, to be more consistent with the error condition name.

**Review by cloud-fan — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
Merged to master, thank you @HyukjinKwon @cloud-fan

**Pull request closed without merging**